File: ensab_ui/src/app/ftree/member-node/member.ts
import { HttpClient } from "@angular/common/http";
import { signal, WritableSignal } from "@angular/core";
import { url } from "../../shared";

export type RawMember = {
  id: string;
  name: string;
  is_male: boolean;
  sons: RawMember[];
};

type SonlessRawMember = {
  id: string;
  name: string;
  is_male: boolean;
};

class Updates {
  origin : RawMember;
  copy : Member;

  constructor(member : Member) {
    this.origin = member.raw()
    this.copy = member;
  }

  private updates(): SonlessRawMember[] {
    const origin = this.origin;
    const copy = this.copy.raw();

    function compare(origin: RawMember, copy: RawMember): SonlessRawMember[] {
        let first : SonlessRawMember | undefined = undefined;
        let rest : SonlessRawMember[] = []
        if (origin.name != copy.name || origin.is_male != copy.is_male) {
            first = {id : copy.id,name : copy.name,is_male : copy.is_male}
        };
        const copy_sons_ids = copy.sons.map(x => x.id);
        const origin_sons = origin
            .sons
            .filter((x) => copy_sons_ids.includes(x.id));
        for (const origin_son of origin_sons) {
          const copy_son = copy.sons.find(x => x.id === origin_son.id)
          if (copy_son) {
            rest = rest.concat(compare(origin_son,copy_son))
          }
        }
        if (first) {
          return [...rest,first]
        } else {
          return rest
        }
    }
    return compare(origin, copy)
  }

  private created():[string,RawMember[]][] {
    const origin = this.origin;
    const copy = this.copy.raw();

    function compare(origin: RawMember, copy: RawMember) : [string,RawMember[]][] {
      const origin_sons_ids = origin.sons.map(x => x.id);
      const copy_sons = copy
          .sons
          .filter(son => !origin_sons_ids.includes(son.id));
      let first : [string,RawMember[]] | undefined= undefined 
      if (copy_sons.length > 0) {
        first = [copy.id,copy_sons]
      }

      let rest : [string,RawMember[]][] = []
      for(const origin_son of origin.sons) {
        const copy_son = copy.sons.find(x => x.id === origin_son.id)
        if (copy_son) {
          rest = rest.concat(compare(origin_son,copy_son))
        }
      }
      if (first) {
        return [...rest,first];
      } else {
        return rest;
      }
    }

    return compare(origin, copy)
  }

  private deleted():string[] {
    const origin = this.origin;
    const copy = this.copy.raw();

    function compare(origin: RawMember, copy: RawMember) : RawMember[] {
      if (!copy || !origin) {
        return []
      }
      const copy_sons_ids = copy.sons.map(x => x.id);
      let first = origin
          .sons
          .filter(son => !copy_sons_ids.includes(son.id))

      let rest : RawMember[] = []
      for(const origin_son of origin.sons) {
        const copy_son = copy.sons.find(x => x.id === origin_son.id)
        if (copy_son) {
          rest = rest.concat(compare(origin_son,copy_son))
        }
      }
      return [...rest,...first];
    }

    return compare(origin, copy).map(x => x.id)
  }


  commit(http: HttpClient) {
    for (const [parent_id, sons] of this.created()) {
      http.post(url(`member/${parent_id}`), sons).subscribe();
    }
    if (this.updates().length > 0) {
      http.put(url("member"), this.updates()).subscribe();
    }
    for (const id of this.deleted()) {
      http.delete(url(`member/${id}`)).subscribe();
    }
    // this.http.post("http://localhost:8080/member", member).subscribe();
  }

}

type Action = 'Preview'|'Add'|'Remove'|'Update'

export default class Member {
  private static instance: Member;
  private static waitlist : WritableSignal<string[]> = signal([])
  static updates : Updates;
  action : WritableSignal<Action>
  id: string;
  name: WritableSignal<string>;
  is_male: WritableSignal<boolean>;
  sons: WritableSignal<Member[]>;

  private constructor(
    name: string,
    id = crypto.randomUUID() as string,
    is_male = true,
    sons = [] as RawMember[],
  ) {
    this.id = id;
    this.name = signal(name);
    this.is_male = signal(is_male);
    this.action = signal('Preview')
    if (sons.length === 0) {
      this.sons = signal([]);
    } else {
      this.sons = signal(
        sons.map((x) => new Member(x.name, x.id, x.is_male, x.sons)),
      );
    }
  }

  takeAction() {
    Member.waitlist.update(xs => [...xs,this.id]);
  }

  checkAction(): boolean {
    const list = Member.waitlist()
    return list.at(list.length -1) === this.id
  }

  private PreviewAction() {
    this.action.set('Preview')
  }

  redrawAction() {
    Member.waitlist.update(xs => xs.filter(x => x !== this.id));
    this.PreviewAction()
  }

  removeAction() {
    this.action.set('Remove')
  }

  updateAction() {
    this.action.set('Update')
  }

  addAction() {
    this.action.set('Add')
  }

  static getInstance(name: string | undefined = undefined): Member {
    if (!this.instance && name) {
      Member.instance = new Member(name);
      Member.updates = new Updates(Member.instance)
    }
    return Member.instance;
  }

  static getInstanceFromRaw({ id, name, is_male, sons }: RawMember): Member {
    Member.instance =  new Member(name, id, is_male, sons);
    Member.updates = new Updates(Member.instance);
    return Member.instance;
  }

  raw(): RawMember {
    const result: RawMember = {
      id: this.id,
      name: this.name(),
      is_male: this.is_male(),
      sons: this.sons().map((x) => x.raw()),
    } as const;
    return result;
  }
  sonless_raw(): SonlessRawMember {
    const result: SonlessRawMember = {
      id: this.id,
      name: this.name(),
      is_male: this.is_male(),
    } as const;
    return result;
  }

  private with_sons(names: string[]): void {
    const name = names.pop();
    if (!name) {
      return;
    }
    const son = new Member(name);
    son.with_sons(names);
    this.sons = signal([son]);
  }

  static create_from_name(name: string): Member {
    const names = name.split("->");
    const namei = names.pop();
    if (!namei) {
      return new Member("");
    }
    const person = new Member(namei);
    person.with_sons(names);
    return person;
  }

  add_son(member : Member) {
    const sons = this.sons()
    const same_person = sons.find(x => x.name() === member.name()) 
    if(same_person){
      for(const person of member.sons()) {
        same_person.add_son(person)
      } 
    } else {
      this.sons.update(xs => [...xs,member])
    }
  }
}

File: ensab/src/family_tree/member.rs
use crate::db::{RawMember, SonlessRawMember};
use leptos::{server, RwSignal, ServerFnError, SignalGetUntracked, SignalSet, SignalUpdate};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Clone, Copy, Default)]
pub enum Action {
    #[default]
    Preview,
    Add,
    Remove,
    Update,
}

#[derive(Clone, Copy, Default)]
pub struct Member {
    pub id: Uuid,
    pub name: RwSignal<String>,
    pub is_male: RwSignal<bool>,
    pub sons: RwSignal<Vec<Member>>,
    pub action: RwSignal<Action>,
}

#[derive(Clone, Copy)]
pub struct Updates {
    origin: RwSignal<RawMember>,
    pub copy: RwSignal<Member>,
}

impl Updates {
    pub fn init(member: Member) -> Self {
        Self {
            origin: RwSignal::from(member.raw()),
            copy: RwSignal::new(member),
        }
    }

    fn updates(&self) -> Vec<SonlessRawMember> {
        let origin = self.origin.get_untracked();
        let copy = self.copy.get_untracked().raw();

        fn compare(origin: RawMember, copy: RawMember) -> Vec<SonlessRawMember> {
            let first = if origin.name != copy.name || origin.is_male != copy.is_male {
                Some(copy.clone().sonless())
            } else {
                None
            };
            let mut rest = Vec::new();
            if let Some(son) = first {
                rest.push(son);
            }
            for origin_son in origin.sons {
                let copy_son = copy.sons.iter().find(|x| x.id == origin_son.id);
                if let Some(copy_son) = copy_son {
                    rest.extend(compare(origin_son, copy_son.clone()));
                }
            }
            rest
        }
        compare(origin, copy)
    }

    fn created(&self) -> Vec<(Uuid, RawMember)> {
        let origin = self.origin.get_untracked();
        let copy = self.copy.get_untracked().raw();

        fn compare(origin: RawMember, copy: RawMember) -> Vec<(Uuid, RawMember)> {
            let mut rest = copy
                .sons
                .iter()
                .filter(|son| origin.sons.iter().all(|x| x.id != son.id))
                .cloned()
                .map(|x| (copy.id, x))
                .collect::<Vec<_>>();
            for origin_son in origin.sons {
                let copy_son = copy.sons.iter().find(|x| x.id == origin_son.id);
                if let Some(copy_son) = copy_son {
                    rest.extend(compare(origin_son, copy_son.clone()));
                }
            }
            rest
        }
        compare(origin, copy)
    }

    fn deleted(&self) -> Vec<Uuid> {
        let origin = self.origin.get_untracked();
        let copy = self.copy.get_untracked().raw();

        fn compare(origin: RawMember, copy: RawMember) -> Vec<RawMember> {
            let mut rest = origin
                .sons
                .iter()
                .filter(|son| copy.sons.iter().all(|x| x.id != son.id))
                .cloned()
                .collect::<Vec<_>>();
            for origin_son in origin.sons {
                let copy_son = copy.sons.iter().find(|x| x.id == origin_son.id);
                if let Some(copy_son) = copy_son {
                    rest.extend(compare(origin_son, copy_son.clone()));
                }
            }
            rest
        }
        compare(origin, copy).into_iter().map(|x| x.id).collect()
    }

    pub async fn commit(&self) -> Result<(), ServerFnError> {
        let updated_members = self.updates();
        let created_members = self.created();
        let deleted_members = self.deleted();
        server_commit(ServerUpdates {
            created_members,
            deleted_members,
            updated_members,
        })
        .await?;
        self.origin.set(self.copy.get_untracked().raw());
        Ok(())
    }

    pub fn discard(&self) {
        self.copy.set(Member::from_raw(self.origin.get_untracked()));
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
struct ServerUpdates {
    created_members: Vec<(Uuid, RawMember)>,
    deleted_members: Vec<Uuid>,
    updated_members: Vec<SonlessRawMember>,
}

#[server(encoding = "Cbor")]
async fn server_commit(updates: ServerUpdates) -> Result<(), ServerFnError> {
    use crate::db::{member, Pool, Postgres};
    use leptos::expect_context;
    let pool = expect_context::<Pool<Postgres>>();
    let mut transaction = pool.begin().await?;
    for (parent_id, member) in updates.created_members {
        let Ok(_) = member::create(&mut transaction, member, Some(parent_id)).await else {
            return Err(ServerFnError::ServerError(
                "error creating member".to_string(),
            ));
        };
    }
    for id in updates.deleted_members {
        let Ok(_) = member::delete(&mut transaction, id).await else {
            return Err(ServerFnError::ServerError(
                "error deleting member".to_string(),
            ));
        };
    }
    let Ok(_) = member::update(&mut transaction, updates.updated_members).await else {
        return Err(ServerFnError::ServerError(
            "error updating member".to_string(),
        ));
    };

    transaction.commit().await?;
    Ok(())
}

pub trait Rm {
    fn find_son(&self, id: Uuid) -> Option<&RawMember>;
    fn sonless(self) -> SonlessRawMember;
}

impl Rm for RawMember {
    fn find_son(&self, id: Uuid) -> Option<&RawMember> {
        if self.id == id {
            return Some(self);
        } else {
            for son in self.sons.iter() {
                if let Some(son) = son.find_son(id) {
                    return Some(son);
                };
            }
        }
        None
    }

    fn sonless(self) -> SonlessRawMember {
        SonlessRawMember {
            id: self.id,
            name: self.name,
            is_male: self.is_male,
        }
    }
}

impl Member {
    pub fn new(name: String) -> Self {
        Self {
            id: Uuid::new_v4(),
            name: RwSignal::new(name),
            is_male: RwSignal::new(true),
            sons: Default::default(),
            action: Default::default(),
        }
    }
    pub fn from_raw(
        RawMember {
            id,
            name,
            is_male,
            sons,
        }: RawMember,
    ) -> Self {
        Self {
            id,
            name: RwSignal::new(name),
            is_male: RwSignal::new(is_male),
            sons: RwSignal::new(sons.into_iter().map(Member::from_raw).collect()),
            action: Default::default(),
        }
    }
    pub fn raw(self) -> RawMember {
        RawMember {
            id: self.id,
            name: self.name.get_untracked(),
            is_male: self.is_male.get_untracked(),
            sons: self
                .sons
                .get_untracked()
                .into_iter()
                .map(|x| x.raw())
                .collect::<Vec<_>>(),
        }
    }
    // pub fn sonless_raw(self) -> SonlessRawMember {
    //     SonlessRawMember {
    //         id: self.id,
    //         name: self.name.get_untracked(),
    //         is_male: self.is_male.get_untracked(),
    //     }
    // }

    fn with_sons(&self, names: &mut Vec<String>) {
        let name = names.pop();
        let Some(name) = name else {
            return;
        };
        let son = Member::new(name);
        son.with_sons(names);
        self.sons.set(vec![son]);
    }

    pub fn create_from_name(name: &str) -> Self {
        let mut names = name.split("->").map(|x| x.to_string()).collect::<Vec<_>>();
        let name = names.pop().unwrap();
        let person = Member::new(name);
        person.with_sons(&mut names);
        person
    }

    pub fn add_son(&self, member: Member) {
        let sons = self.sons.get_untracked();
        let same_person = sons
            .iter()
            .find(|x| x.name.get_untracked() == member.name.get_untracked());
        match same_person {
            Some(same_person) => {
                for person in member.sons.get_untracked() {
                    same_person.add_son(person)
                }
            }
            None => {
                self.sons.update(|xs| xs.push(member));
            }
        }
    }
}
